### PR TITLE
Treat ASCII-only literals as str when using unicode_literals

### LINKED ISSUE #3619: Treat ascii-only literals as str, not unicode when using `unicode_literals`
Many typeshed stubs only accept `str` when ascii-only `unicode` values also work at runtime. This causes a lot of problems with code using `unicode_literals`, since generally most literals will have `unicode` as their type. A quick fix would be to infer `str` as the type for ascii-only string literals when using `unicode_literals`, but only then. This would cause some unsafety, but it may be worth it for users using `unicode_literals`. The type safe solution is to manually check every `str` argument type in the entire typeshed and potentially replace it with `unicode` or similar, but this would be extremely painful and unlikely to happen any time soon. This would have no effect on code that doesn't use `unicode_literals`.

Example of how this would work:

```py
from __future__ import unicode_literals

reveal_type('foo')   # str
reveal_type('ß')   # unicode
```

Some open questions:
* Should `u'foo'` always have type `unicode`?
* What's the inferred type of `x` in `x = 'foo'`?

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes #3619 

This PR only affects code that uses ``from __future__ import unicode_literals``.

It looks like there is a solution that will not require modifying ``typed_ast``. Note that this will treat ASCII-only literals with an explicit prefix ``u'foo'`` as ``str``, not ``unicode``. This can be changed, but only by updating ``typed_ast`` (the point is that ``Str`` has an attribute ``has_b`` indicating the presence of ``b`` prefix, we will need also ``has_u``).

I know ``unicode_literals`` are bad, but I have seen several people complaining about this recently.
@JukkaL @JelleZijlstra what do you think?

### CODE REVIEW

File: mypy/fastparse2.py [hunk 1]
```diff
@@ -637,6 +638,9 @@ def visit_ImportFrom(self, n: ast27.ImportFrom) -> ImportBase:
                            n.level,
                            [(a.name, a.asname) for a in n.names])
         self.imports.append(i)
+        if (n.module == '__future__' and len(n.names) == 1 and
```
Comment (JelleZijlstra) [inline mypy/fastparse2.py]: Will this work if I do `from __future__ import absolute_import, unicode_literals`? That works at runtime.

Also, I don't think the `asname` check is necessary. This program:
```
from __future__ import unicode_literals as why_would_you_do_this
print(type(''))
```
prints "unicode".

Comment (ilevkivskyi) [inline mypy/fastparse2.py]: OK, good points. I will relax the check now.

**Comment by JelleZijlstra** [comment]
I left some comments on the implementation. If it's not too hard to do in typed_ast, I think we should also make sure that strings with an explicit `u` prefix remain unicode.

I'm also not entirely sure this is a good idea, since it does introduce a difference with how these things actually work at runtime. For example, perhaps somebody is passing `['a', 'list', 'of', 'strings']` to a function that accepts `List[unicode]`, and with this change the list will become a `List[str]` instead of `List[unicode]`. We should see if that is an issue in real-world codebases that use `unicode_literals`; @rowillia could you help check that?

File: test-data/unit/python2eval.test [hunk 2]
```diff
@@ -151,15 +151,27 @@ f(**params)
...
+f(u'щось')
+f('щось інше')
 [out]
-_program.py:5: error: Argument 1 to "f" has incompatible type "unicode"; expected "str"
 _program.py:6: error: Argument 1 to "f" has incompatible type "unicode"; expected "str"
+_program.py:7: error: Argument 1 to "f" has incompatible type "unicode"; expected "str"
+
+[case testUnicodeLiteralsFutureImportAllASCII_python2]
+# coding=utf-8
+from __future__ import unicode_literals
```
Comment (JelleZijlstra) [inline test-data/unit/python2eval.test]: Could you add tests for some of the other syntactic variations, like `from __future__ import absolute_import, unicode_literals` and `from __future__ import unicode_literals as something_else`?

Comment (ilevkivskyi) [inline test-data/unit/python2eval.test]: OK, will do.

**Comment by ilevkivskyi** [comment]
@JelleZijlstra 
> If it's not too hard to do in typed_ast, I think we should also make sure that strings with an explicit ``u`` prefix remain unicode.

This is basically one-two lines (plus some boilerplate, but it is auto-generated). The main problem is that this will require a ``typed_ast`` release and updating its stub. I can do this if we agree we need this change.

**Comment by ilevkivskyi** [comment]
Are we going to push in this direction? If it is not that important, maybe better just close this?

**Comment by JukkaL** [comment]
This requires more analysis/experimentation to proceed. Also, I think that we'd need to handle the explicit `u` prefix.

**Comment by ilevkivskyi** [comment]
> Also, I think that we'd need to handle the explicit ``u`` prefix.

OK, I will then prepare a ``typed_ast`` PR (this PR can be updated however only after the release of ``typed_ast``, otherwise tests will fail).

**Comment by ilevkivskyi** [comment]
The corresponding ``typed_ast` PR https://github.com/python/typed_ast/pull/49

**Comment by ilevkivskyi** [comment]
I am closing this since it seems many people don't like `unicode_literals`. Instead I propose to have a separate discussion about how to use the `typed_ast` feature we are not using: `typed_ast` preserves original prefix for string/bytes literals in both Python 2 and 3.

**Comment by gvanrossum** [comment]
Is that separate discussion somewhere? Because in the end we did release the typed_ast feature (with a lot of delays and some fumblings). Was it all for naught? Or is there actually some use for the `kind` attribute on `Str`?

**Comment by gvanrossum** [comment]
Oh, that's a good enough use case for me!

**Pull request closed without merging**